pkg/distributor/distributor.go

```diff
@@ -483,18 +497,35 @@ func (d *Distributor) validateSeries(ts cortexpb.PreallocTimeseries, userID stri
 		return emptyPreallocSeries, err
 	}
 
-	samples := make([]cortexpb.Sample, 0, len(ts.Samples))
-	for _, s := range ts.Samples {
-		if err := validation.ValidateSample(d.limits, userID, ts.Labels, s); err != nil {
-			return emptyPreallocSeries, err
+	// Don't alloc a new slice unnecessarily when input is empty
+	samples := ts.Samples
+	if len(ts.Samples) > 0 {
+		samples = make([]cortexpb.Sample, 0, len(ts.Samples))
+		for _, s := range ts.Samples {
+			if err := validation.ValidateSample(d.limits, userID, ts.Labels, s); err != nil {
+				return emptyPreallocSeries, err
+			}
+			samples = append(samples, s)
+		}
+	}
+
+	// Don't alloc a new slice unnecessarily when input is empty
+	exemplars := ts.Exemplars
+	if len(ts.Exemplars) > 0 {
+		exemplars = make([]cortexpb.Exemplar, 0, len(ts.Exemplars))
+		for _, e := range ts.Exemplars {
+			if err := validation.ValidateExemplar(userID, ts.Labels, e); err != nil {
+				return emptyPreallocSeries, err
```

Comment:
In theory yes, a `TimeSeries` can contain both samples and exemplars, but in the current implementation of remote write it would only contain one or the other.

Reply:
> but in the current implementation of remote write it would only contain one or the other.

I would add a comment about it. I think it's important to mention because in case they will contain both, then return error is may be wrong. We don't want to skip ingesting samples just because exemplars are invalid.